abcix/adapter/adapter.go

```diff
@@ -19,59 +22,73 @@ func (app *adaptedApp) OriginalApp() abci.Application {
 	return app.abciApp
 }
 
-func (app *adaptedApp) Info(req abcix.RequestInfo) (resp abcix.ResponseInfo) {
-	abciReq := abci.RequestInfo{}
-	if err := copier.Copy(&abciReq, &req); err != nil {
-		// TODO: panic for debugging purposes. better error handling soon!
-		panic(err)
+var typeRegistry = map[string]reflect.Type{
+	"abci.RequestInfo":               reflect.TypeOf(abci.RequestInfo{}),
+	"abci.RequestSetOption":          reflect.TypeOf(abci.RequestSetOption{}),
+	"abci.RequestQuery":              reflect.TypeOf(abci.RequestQuery{}),
+	"abci.RequestCheckTx":            reflect.TypeOf(abci.RequestCheckTx{}),
+	"abci.RequestInitChain":          reflect.TypeOf(abci.RequestInitChain{}),
+	"abci.RequestListSnapshots":      reflect.TypeOf(abci.RequestListSnapshots{}),
+	"abci.RequestOfferSnapshot":      reflect.TypeOf(abci.RequestOfferSnapshot{}),
+	"abci.RequestLoadSnapshotChunk":  reflect.TypeOf(abci.RequestOfferSnapshot{}),
+	"abci.RequestApplySnapshotChunk": reflect.TypeOf(abci.RequestApplySnapshotChunk{}),
+	"abci.RequestEndBlock":           reflect.TypeOf(abci.RequestEndBlock{}),
+	"abci.RequestBeginBlock":         reflect.TypeOf(abci.RequestBeginBlock{}),
+}
+
+func newStruct(name string) (interface{}, bool) {
+	elem, ok := typeRegistry[name]
+	if !ok {
+		return nil, false
 	}
-	abciResp := app.abciApp.Info(abciReq)
-	if err := copier.Copy(&resp, &abciResp); err != nil {
-		// TODO: panic for debugging purposes. better error handling soon!
-		panic(err)
+	return reflect.New(elem).Interface(), true
+}
+
+func Apply(f interface{}, args ...interface{}) reflect.Value {
+	fun := reflect.ValueOf(f)
+	in := make([]reflect.Value, len(args))
+	for k, arg := range args {
+		in[k] = reflect.ValueOf(arg).Elem()
 	}
-	return
+	r := fun.Call(in)
+	return r[0]
 }
 
-func (app *adaptedApp) SetOption(req abcix.RequestSetOption) (resp abcix.ResponseSetOption) {
-	abciReq := abci.RequestSetOption{}
-	if err := copier.Copy(&abciReq, &req); err != nil {
+func adaptGeneralization(req interface{}, resp interface{}, abciReq string, f interface{}) error {
+	abcireq, ok := newStruct(abciReq)
+	if !ok {
+		return errors.New("fail")
+	}
+	if err := copier.Copy(abcireq, req); err != nil {
 		// TODO: panic for debugging purposes. better error handling soon!
 		panic(err)
 	}
-	abciResp := app.abciApp.SetOption(abciReq)
-	if err := copier.Copy(&resp, &abciResp); err != nil {
+
+	abciResp := Apply(f, abcireq).Interface()
+	if err := copier.Copy(resp, abciResp); err != nil {
 		// TODO: panic for debugging purposes. better error handling soon!
 		panic(err)
 	}
+	return nil
+}
+
+func (app *adaptedApp) Info(req abcix.RequestInfo) (resp abcix.ResponseInfo) {
+	adaptGeneralization(&req, &resp, "abci.RequestInfo", app.abciApp.Info)
```

Comment:
instead of the third argument as a string, can you just use the function name? 

https://stackoverflow.com/a/7053871

Reply:
The third argument is used to map the struct name into the desired struct, the fourth argument is the function name.